app/cloud.go

```diff
@@ -4,14 +4,105 @@
 package app
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"time"
 
+	"github.com/mattermost/mattermost-server/v6/app/request"
 	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/i18n"
 	"github.com/mattermost/mattermost-server/v6/shared/mlog"
 )
 
+func (a *App) NotifySystemAdminsToUpgrade(c *request.Context, currentUserTeamID string) *model.AppError {
+	userId := c.Session().Id
```

Comment:
This is better being in the app layer! We only want one section of locking & unlocking though. Otherwise, arbitrarily many requests could read at the same time, and then whichever is the last to write will clobber the others, losing track of their info. Here's an example of what I mean. If you run this with `go run temp.go`, you will sometimes get different amounts of elements in different runs.

temp.go
```go
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	mu := sync.Mutex{}
	wg := sync.WaitGroup{}
	wg.Add(2)

	sharedData := []int{}

	go func() {
		for i := 0; i < 1000; i += 2 {
			// read
			mu.Lock()
			val := sharedData
			mu.Unlock()

			// do other works/checks
			time.Sleep(time.Nanosecond)
			newVal := append(val, i)
			fmt.Printf("goroutine 2 %v\n", val)

			// write
			mu.Lock()
			sharedData = newVal
			mu.Unlock()
		}
		wg.Done()
	}()

	go func() {
		for i := 1; i < 1000; i += 2 {
			// read
			mu.Lock()
			val := sharedData
			mu.Unlock()

			// do other works/checks
			time.Sleep(time.Nanosecond)
			newVal := append(val, i)
			fmt.Printf("goroutine 2 %v\n", val)

			// write
			mu.Lock()
			sharedData = newVal
			mu.Unlock()
		}
		wg.Done()
	}()
	wg.Wait()
	fmt.Printf("final result %v, %d elements\n", sharedData, len(sharedData))
}
```

Reply:
Made a PR that changes it to one critical section: https://github.com/mattermost/mattermost-server/pull/20488/files , and makes sending the notifications async since we don't care about the result of that.